src/libstd/panicking.rs

```diff
@@ -222,19 +222,65 @@ fn default_hook(info: &PanicInfo<'_>) {
 #[cfg(not(test))]
 #[doc(hidden)]
 #[unstable(feature = "update_panic_count", issue = "none")]
-pub fn update_panic_count(amt: isize) -> usize {
+pub mod panic_count {
     use crate::cell::Cell;
-    thread_local! { static PANIC_COUNT: Cell<usize> = Cell::new(0) }
+    use crate::sync::atomic::{AtomicUsize, Ordering};
+
+    // Panic count for the current thread.
+    thread_local! { static LOCAL_PANIC_COUNT: Cell<usize> = Cell::new(0) }
+
+    // Sum of panic counts from all threads. The purpose of this is to have
+    // a fast path in `is_zero` (which is used by `panicking`). Access to
+    // this variable can be always be done with relaxed ordering because
+    // it is always guaranteed that, if `GLOBAL_PANIC_COUNT` is zero,
+    // `LOCAL_PANIC_COUNT` will be zero.
```

Comment:
> if `GLOBAL_PANIC_COUNT` is zero, `LOCAL_PANIC_COUNT` will be zero.

This is a very strong statement to make, in particular about relaxed atomics! It took me some time to actually follow the argument here. You cannot, in general, relate the content of two *different* locations when you use relaxed accesses. In particular, "`GLOBAL_PANIC_COUNT` is zero" is an odd statement as atomics can appear to have different values in different threads. The only reason it works here is that one of the two is thread-local.  I was also confused at first because `decrease` decrements `GLOBAL` first, so there is a time when `GLOBAL` is already zero for the current thread but `LOCAL` is still 1.

So what about

> In any particular thread, if *that thread* currently views `GLOBAL_PANIC_COUNT` as being zero, then `LOCAL_PANIC_COUNT` *in that thread* is zero. This only holds *outside* of `increase` and `decrease`.

Also, it would be good to explain in a comment why accessing an atomic global (potentially subject to contention)  is faster than accessing a non-atomic thread local. That is rather non-intuitive IMO -- I would have expected the exact opposite.

Reply:
Getting a TLS pointer requires calling `__tls_get_addr` for ELF binaries when using the GD TLS model. This is not inlinable and may walk a linked list.

https://github.com/bminor/glibc/blob/5f72f9800b250410cad3abfeeb09469ef12b2438/elf/dl-tls.c#L821

Reply:
> In any particular thread, if _that thread_ currently views `GLOBAL_PANIC_COUNT` as being zero, then `LOCAL_PANIC_COUNT` _in that thread_ is zero. This invariant holds _before_ and _after_ `increase` and `decrease`, but not necessarily during their execution.

Yes, that would have been a better description. I wrote the comment thinking about the `is_zero` function, which is the only place where the value of `GLOBAL_PANIC_COUNT` matters.

> Also, it would be good to explain in a comment why accessing an atomic global (potentially subject to contention) is faster than accessing a non-atomic thread local. That is rather non-intuitive IMO -- I would have expected the exact opposite.

You can compare [here](https://godbolt.org/z/EoppL-) the assembly of a relaxed atomic load and a TLS read. The atomic load just needs a normal `mov`.

Reply:
> You can compare here the assembly of a relaxed atomic load and a TLS read. The atomic load just needs a normal mov.

Assembly is basically gibberish to me.^^ But that's okay, a high-level explanation like @bjorn3 gave is totally sufficient. :)

So could you open a PR to include these comments, or should I do that?

Reply:
> So could you open a PR to include these comments, or should I do that?

I'll do it now.